lib/elixir/src/elixir_clauses.erl

```diff
@@ -169,11 +169,16 @@ with(Meta, Args, S, E) ->
 
   {{with, Meta, EExprs ++ [[{do, EDo} | EOpts]]}, S3, E}.
 
-expand_with({'<-', Meta, [Left, Right]}, {S, E, _HasMatch}) ->
+expand_with({'<-', Meta, [Left, Right]}, {S, E, HasMatch}) ->
   {ERight, SR, ER} = elixir_expand:expand(Right, S, E),
   SM = elixir_env:reset_read(SR, S),
   {[ELeft], SL, EL} = head([Left], SM, ER),
-  {{'<-', Meta, [ELeft, ERight]}, {SL, EL, true}};
+  case ELeft of
+    {Var, _, Ctx} when is_atom(Var), is_atom(Ctx) ->
+      {{'=', Meta, [ELeft, ERight]}, {SL, EL, HasMatch}};
```

Comment:
This is nice, it is even able to emit a warning in this case

<img width="671" alt="Screenshot 2024-01-30 at 19 45 02" src="https://github.com/elixir-lang/elixir/assets/11598866/08eb883f-2d60-41bc-92b3-069cb12339ad">

Reply:
fantastic

Reply:
However, we should probably do the conversion when translating Erlang code, to avoid doing many changes to the original AST. The original AST is used for type checking and in error reports, so we want to preserve semantics when we can.

Reply:
Makes sense! Addressed in https://github.com/elixir-lang/elixir/pull/13299/commits/4e11ff285b26539874e640bd6a4620d03d9789ed